Improve 'all' tab pagination to handle edge cases

jazairi opened the pull request:
## NOTE: Changes during code review have drifted this PR message out of sync. See individual commit messages for changelog.

#### Why these changes are being introduced:

The zipper merge we implemented naively queries n/2 results from each API and interleaves them, where n is the per-page value. This works if both APIs return many results, but it can cause problems in smaller, unbalanced result sets.

For example, the query term `doc edgerton` returns 50 Primo results and 4 TIMDEX results. Page 1 only shows 14 results (4 TIMDEX and 10 Primo), and each subsequent page returns only 10 (all Primo).

#### Relevant ticket(s):

- [USE-179](https://mitlibraries.atlassian.net/browse/USE-179)

#### How this addresses that need:

This implements more sophisticated logic that first checks the number of hits returned by each API and passes that, along with the pagination information, to a Merged Search Paginator class. This service object develops a 'merge plan', calculates API offsets, and merges the results for each page.

Queries on the 'all' tab now fetch twice from each API: once to determine the total number of hits for the Merged Search Paginator then again to fetch results at the appropriate offset. While hardly ideal, this was the only option I could figure to avoid losing results. I limited these extra calls to queries beyond page 1, which is the only case where they are needed.

#### Side effects of this change:

* We now clear cache before each search controller test. This was done to avoid odd test behavior, but I ran the suite 50 times without any issues, so it might be excessively cautious.
* The search controller continues to grow with this new logic. I tried to split things into multiple helper methods, so if we want to move more things to service objects later, it might be easier to do so.
* A failing cassette has been replaced with a mock.

#### Developer

##### Accessibility

- [ ] ANDI or WAVE has been run in accordance to [our guide](https://mitlibraries.github.io/guides/basics/a11y.html).
- [x] This PR contains no changes to the view layer.
- [ ] New issues flagged by ANDI or WAVE have been resolved.
- [ ] New issues flagged by ANDI or WAVE have been ticketed (link in the Pull Request details above).
- [ ] No new accessibility issues have been flagged.

##### New ENV

- [ ] All new ENV is documented in README.
- [ ] All new ENV has been added to Heroku Pipeline, Staging and Prod.
- [x] ENV has not changed.

##### Approval beyond code review

- [ ] UXWS/stakeholder approval has been confirmed.
- [x] UXWS/stakeholder review will be completed retroactively.
- [ ] UXWS/stakeholder review is not needed.

##### Additional context needed to review

This is a pretty unwieldy changeset, so please reach out if you have questions!

#### Code Reviewer

##### Code

- [ ] I have confirmed that the code works as intended.
- [ ] Any CodeClimate issues have been fixed or confirmed as
added technical debt.

##### Documentation

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message).
- [ ] The documentation has been updated or is unnecessary.
- [ ] New dependencies are appropriate or there were no changes.

##### Testing

- [ ] There are appropriate tests covering any new functionality.
- [ ] No additional test coverage is required.


[USE-179]: https://mitlibraries.atlassian.net/browse/USE-179?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jazairi committed:
Improve 'all' tab pagination to handle edge cases

Why these changes are being introduced:

The zipper merge we implemented naively queries n/2 results from each
API and interleaves them, where n is the per-page value. This works if
both APIs return many results, but it can cause problems in smaller,
unbalanced result sets.

For example, the query term `doc edgerton` returns 50 Primo results and
4 TIMDEX results. Page 1 only shows 14 results (4 TIMDEX and 10 Primo),
and each subsequent page returns only 10 (all Primo).

Relevant ticket(s):

- [USE-179](https://mitlibraries.atlassian.net/browse/USE-179)

How this addresses that need:

This implements more sophisticated logic that first checks the number
of hits returned by each API and passes that, along with the pagination
information, to a Merged Search Paginator class. This service object
develops a 'merge plan', calculates API offsets, and merges the results
for each page.

Queries on the 'all' tab now fetch twice from each API: once to
determine the total number of hits for the Merged Search Paginator
then again to fetch results at the appropriate offset. While hardly
ideal, this was the only option I could figure to avoid losing results.
I limited these extra calls to queries beyond page 1, which is the
only case where they are needed.

Side effects of this change:

* We now clear cache before each search controller test. This was done
to avoid odd test behavior, but I ran the suite 50 times without any
issues, so it might be excessively cautious.
* The search controller continues to grow with this new logic. I tried
to split things into multiple helper methods, so if we want to move
more things to service objects later, it might be easier to do so.
* A failing cassette has been replaced with a mock.

jazairi committed:
Refactor all tab logic to service and enable cache

Why these changes are being introduced:

In discussions of the PR in review for USE-179, we determined that the
proposed pagination improvements could be more efficient. We had also
determined that the code changes were difficult to follow, and could
use better documentation.

Relevant ticket(s):

- USE-179

How this addresses that need:

This commit caches the page 1 'summary' API calls, which we use to
gather the hit counts from each API to calculate pagination on deeper
pages. It also abstracts the 'all' tab code to a 'Merged Search Service',
mirroring the design pattern of the Merged Search Paginator, and adds
docstrings to the methods in that service.

Side effects of this change:

We are still making two API calls on deeper pages when the hit totals
are not cached. I could not find a workaround to this while still
supporting nonlinear pagination. However, the vast majority of users
(even bots, presumably) will begin their search at page 1, so hopefully
this is a rare occurrence.

JPrevost commented on app/models/merged_search_service.rb:
```diff
@@ -0,0 +1,241 @@
+    @enhanced_query = enhanced_query
+    @active_tab = active_tab
+    @cache = cache
+    @primo_fetcher = primo_fetcher || method(:default_primo_fetch)
```
It's unclear when/why we are injecting versus using the default methods.

The controller seems to be injecting so I'm not sure when the defaults are used.

JPrevost replied:
It's still unclear to me if we ever use the default path or if we are always injecting... it seems like the Controller and Tests both inject so is default never used?

JPrevost commented on app/models/merged_search_paginator.rb:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  # Merges two result arrays according to the merge plan
+  def merge_results(primo_results, timdex_results)
```
This feels more like part of the merge service than the paginator service.

jazairi replied:
Probably. It started in the paginator service because it predated the merge service, and I left it there out of laziness. I agree that it makes more sense in the merge service, though. I'll explore what that might entail (I think it's just rewriting some tests).

JPrevost commented on app/models/merged_search_service.rb:
```diff
@@ -0,0 +1,241 @@
+
+    merged = paginator.merge_results(primo_data[:results] || [], timdex_data[:results] || [])
+    errors = combine_errors(primo_data[:errors], timdex_data[:errors])
+    pagination = Analyzer.new(@enhanced_query, timdex_total, :all, primo_total).pagination
```
I'm not sure if this is a bug that can happen in this application, or just in the console.

But here's a sequence of console commands that lead to an unexpected result for this pagination value (tldr; it uses the per_page for the app whereas all of these methods allow you to pass any value you want... maybe they shouldn't accept values and just use the per_page of the app to avoid having someone override something at some point and have pagination blow up?)

```
params = {q: 'hello'}
q = Enhanger.new(params)
s = MergedSearchService.new(enhanced_query: q, active_tab: 'all')
s.fetch(page: 1, per_page: 1)
```

Pagination will then be:
```
pagination: {hits: 162397, per_page: 20, start: 1, end: 20, next: 2}
```

Which has the correct offsets and next, but the wrong per_page.

JPrevost commented on test/models/merged_search_service_test.rb:
```diff
@@ -0,0 +1,202 @@
+require 'test_helper'
+require 'ostruct'
```
Are we only using OpenStruct in testing or has it ended up on the running app as well?

Zero concerns if it's just in tests, minor nervousness if it is in the main codebase because it has the _potential_ to cause unexpected slow downs. Mostly we'd just want to keep an eye on it, not avoid its use.

jazairi replied:
We're just using it in testing.

JPrevost reviewed (commented):
I still need to work through the test suites for the two new models but wanted to share my feedback so far as I need to step away from the review for a bit and thought in-progress feedback may be helpful rather than waiting until I can get back to it.

jazairi commented on app/models/merged_search_service.rb:
```diff
@@ -0,0 +1,235 @@
+  # discoverability for stores that are probed with `read_matched`.
+  #
+  # @param totals [Hash] { primo: Integer, timdex: Integer }
+  def write_cached_totals(totals)
```
Honestly, this method can probably be simplified a lot. For example, one cache write was for debugging while I was writing the feature, and I thought it might be useful for future debugging. Looking at this with fresh eyes, though, I don't think it's necessary..

I've got a version that it is hopefully clearer on intent, which I'll push up once I've gone through the remaining feedback.

jazairi commented on app/models/merged_search_service.rb:
```diff
@@ -0,0 +1,235 @@
+
+      results = assemble_all_tab_result(paginator, primo_data, timdex_data, current_page, per_page)
+
+      return results
```
If you think it would improve readability, that's good enough for me!

jazairi committed:
Address code review feedback

- Improve documentation
- Adjust TTL
- Remove incorrect Primo continuation calculation
- Move merge_results method to Merged Search
service
- Add optional per_page param to Analyzer's
initialize method

JPrevost commented on app/models/merged_search_service.rb:
```diff
@@ -0,0 +1,319 @@
+  # calling code simple and to make testing deterministic.
+  #
+  # Usage patterns:
+  # - Default: callers omit fetcher arguments and the service uses `Rails.cache`
```
It's still a bit unclear to me if we ever use the default path or if we are always injecting...

JPrevost commented on test/test_helper.rb:
```diff
@@ -42,3 +42,18 @@ class TestCase
+#
+# If you hit flaky tests related to cache  access under concurrency, consider injecting a per-test
+# cache or avoiding shared mutable state in the threaded path.
+Rails.cache = ActiveSupport::Cache::MemoryStore.new
```
It might be better to change this value in the config rather than overriding it here?

https://github.com/MITLibraries/timdex-ui/blob/main/config/environments/test.rb#L26

jazairi committed:
Address code review feedback

- Reduce duplication for per-page logic in analyzer
- Move cache assignment from test helper to test environment config
- Remove default fetch methods and add test factory to inject fetchers
for relevant tests

JPrevost reviewed (approved):
:shipit:
